nano/node/confirmation_height_processor.cpp

```diff
@@ -0,0 +1,294 @@
+#include <boost/optional.hpp>
+#include <boost/polymorphic_pointer_cast.hpp>
+#include <nano/lib/numbers.hpp>
+#include <nano/lib/timer.hpp>
+#include <nano/lib/utility.hpp>
+#include <nano/node/confirmation_height_processor.hpp>
+#include <nano/node/node.hpp> // For active_transactions
+#include <nano/secure/blockstore.hpp>
+#include <nano/secure/common.hpp>
+#include <nano/secure/ledger.hpp>
+
+constexpr std::chrono::milliseconds nano::confirmation_height_processor::batch_write_delta;
+
+nano::confirmation_height_processor::confirmation_height_processor (nano::pending_confirmation_height & pending_confirmation_height, nano::block_store & store, nano::ledger & ledger, nano::active_transactions & active, nano::logger_mt & logger) :
+pending_confirmations (pending_confirmation_height),
+store (store),
+ledger (ledger),
+active (active),
+logger (logger),
+thread ([this]() {
+	nano::thread_role::set (nano::thread_role::name::confirmation_height_processing);
+	this->run ();
+})
+{
+}
+
+nano::confirmation_height_processor::~confirmation_height_processor ()
+{
+	stop ();
+}
+
+void nano::confirmation_height_processor::stop ()
+{
+	stopped = true;
+	condition.notify_one ();
+	if (thread.joinable ())
+	{
+		thread.join ();
+	}
+}
+
+void nano::confirmation_height_processor::run ()
+{
+	std::unique_lock<std::mutex> lk (pending_confirmations.mutex);
+	while (!stopped)
+	{
+		if (!pending_confirmations.queue.empty ())
+		{
+			auto pending_confirmation = pending_confirmations.queue.front ();
+			pending_confirmations.queue.pop ();
+			lk.unlock ();
+			add_confirmation_height (pending_confirmation);
+			lk.lock ();
```

Comment:
We can externally signal confirmation here if the block was marked as live when it was inserted in to pending_confirmations.queue.

Reply:
If the plan is to manually trigger callback and add record in confirmation history for external signal, updates here may be useful for reference: https://github.com/nanocurrency/nano-node/pull/1898/files. The block_info/blocks_info RPC calls also need context of which blocks are confirmed during upgrade (since they return a "confirmed" bool), is that already covered by a check for the requested block in the pending_confirmations.queue?

Reply:
Is this still needed with the approach we have now taken? (i.e to know when that block's confirmation height has been finalised). If so perhaps can wait until RC2.

Reply:
I believe this should be covered by the approach we have now taken with block_info/blocks_info providing confirmation context for the requested block(s) only.

Reply:
Thought so too, thanks :)